### PR TITLE
Sb 13896

### PR DESCRIPTION
**Pull request by Hari-stackroute**
merge user account changes

### CODE REVIEW

File: actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java [hunk 1]
```diff
@@ -40,6 +40,7 @@
 @ActorConfig(
   tasks = {},
   asyncTasks = {
+          "mergerUserSToElastic",
```
Comment (manzarul) [inline actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java]: mergerUserSToElastic to mergerUserToElastic

File: actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java [hunk 2]
```diff
@@ -53,7 +54,7 @@
     "updateUserNotesToElastic",
     "insertUserCoursesInfoToElastic",
     "updateCourseBatchToEs",
-    "insertCourseBatchToEs"
+    "insertCourseBatchToEs",
```
Comment (manzarul) [inline actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java]: remove comman  from insertCourseBatchToEs .

File: actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java [hunk 3]
```diff
@@ -113,6 +114,8 @@ public void onReceive(Request request) throws Throwable {
       insertUserNotesToEs(request);
     } else if (operation.equalsIgnoreCase(ActorOperations.UPDATE_USER_NOTES_ES.getValue())) {
       updateUserNotesToEs(request);
+    } else if (operation.equalsIgnoreCase(ActorOperations.MERGE_USERS_TO_ELASTIC.getValue())) {
```
Comment (manzarul) [inline actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java]: MERGE_USERS_TO_ELASTIC to MERGE_USER_TO_ELASTIC

File: actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java [hunk 4]
```diff
@@ -628,4 +631,13 @@ private void updateUserNotesToEs(Request actorMessage) {
         (String) noteMap.get(JsonKey.ID),
         noteMap);
   }
+
+  private void mergeUserDetailsToEs(Request mergeRequest) {
+    String mergeeId = (String) mergeRequest.get("mergeeId");
```
Comment (manzarul) [inline actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java]: use same termonologu either user fromAccount and toAccount or mergeeId or mergerId

File: actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java [hunk 5]
```diff
@@ -628,4 +631,13 @@ private void updateUserNotesToEs(Request actorMessage) {
...
+    Map<String, Object> mergeeMap = (Map<String, Object>) mergeRequest.get("userFromAccount");
+    updateDataToElastic(ProjectUtil.EsIndex.sunbird.getIndexName(),
```
Comment (manzarul) [inline actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java]: Add some logs.

File: actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+)
+public class UserMergeActor extends UserBaseActor {
+    private ObjectMapper objectMapper = new ObjectMapper();
+    private UserService userService = UserServiceImpl.getInstance();
+    @Override
+    public void onReceive(Request userRequest) throws Throwable {
+        updateUserMergeDetails(userRequest);
+    }
+
+    private void updateUserMergeDetails(Request userRequest) throws IOException {
```
Comment (manzarul) [inline actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java]: Please split it into different private method.

File: actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+                            ResponseCode.mergeeIdNotExists.getErrorMessage(), mergeeId),
+                    ResponseCode.CLIENT_ERROR.getResponseCode());
+        }
+
+
+    }
+
+    private void mergeUserDetailsToEs(Request userRequest) {
+        userRequest.setOperation(ActorOperations.MERGE_USERS_TO_ELASTIC.getValue());
+        ProjectLogger.log(String.format("%s:%s:Trigger sync of user details to ES with user updated userMap %s", this.getClass().getSimpleName(), "mergeUserDetailsToEs", Collections.singleton(userRequest.toString())));
```
Comment (manzarul) [inline actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java]: no need to print complete data.
add loger as info here

File: actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        tellToAnother(userRequest);
+    }
+
+    private String updateMergerCourseDetails(Map<String, Object> requestMap) throws IOException {
+        //call course service api
+        String bodyJson = objectMapper.writeValueAsString(requestMap);
+        Map headersMap = new HashMap();
+        String responseCode = null;
+        //courseurl need to provide
```
Comment (manzarul) [inline actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java]: Remove unused code

File: actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+            mergeeESMap.put(JsonKey.STATUS, 0);
+            mergeeESMap.put(JsonKey.IS_DELETED, true);
+            mergeeESMap.put(JsonKey.EMAIL,null);
+            mergeeESMap.put(JsonKey.PHONE,null);
+            mergeeESMap.put(JsonKey.USERNAME, null);
+            mergeeESMap.put(JsonKey.PREV_USED_EMAIL, mergee.getEmail());
+            mergeeESMap.put(JsonKey.PREV_USED_PHONE, mergee.getPhone());
+            mergeeESMap.put(JsonKey.UPDATED_DATE, ProjectUtil.getFormattedDate());
+            mergeeESMap.put(JsonKey.USER_ID, mergeeId);
```
Comment (amit-tarento) [inline actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java]: both mergeeDBMap and mergeeESMap have same set of data so just create one and use map.putAll

File: actors/user/src/main/java/org/sunbird/user/dao/impl/UserDaoImpl.java [hunk 10]
```diff
@@ -55,6 +55,11 @@ public Response updateUser(User user) {
     return cassandraOperation.updateRecord(Util.KEY_SPACE_NAME, TABLE_NAME, map);
   }
 
+  @Override
+  public Response updateUserFieldsWithNullValues(Map<String, Object> userMap) {
```
Comment (amit-tarento) [inline actors/user/src/main/java/org/sunbird/user/dao/impl/UserDaoImpl.java]: just overload public Response updateUser(User user) method which will accept Map<String, Object>

File: actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java (hunk 1 above)
Comment (Hari-stackroute) [reply actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java]: done

File: actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java (hunk 2 above)
Comment (Hari-stackroute) [reply actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java]: done

File: actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java (hunk 3 above)
Comment (Hari-stackroute) [reply actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java]: Done

File: actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java (hunk 4 above)
Comment (Hari-stackroute) [reply actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java]: followed fromAccountId and toAccountId while getting from request.

File: actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java (hunk 5 above)
Comment (Hari-stackroute) [reply actors/common/src/main/java/org/sunbird/learner/actors/BackgroundJobManager.java]: Added

File: actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java (hunk 8 above)
Comment (Hari-stackroute) [reply actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java]: This code for course call, so commented as of now.

File: actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java (hunk 9 above)
Comment (Hari-stackroute) [reply actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java]: yes code is redundant here, removed mergeeESMap completely and used the above one for cassandra and ES calls

File: actors/user/src/main/java/org/sunbird/user/dao/impl/UserDaoImpl.java (hunk 10 above)
Comment (Hari-stackroute) [reply actors/user/src/main/java/org/sunbird/user/dao/impl/UserDaoImpl.java]: Done

File: actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java (hunk 6 above)
Comment (Hari-stackroute) [reply actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java]: Now the method size is shredded.

File: actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java (hunk 7 above)
Comment (Hari-stackroute) [reply actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java]: Done

File: actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+                LoggerEnum.DEBUG.name());
+        Map<String, Object> targetObject = null;
+        List<Map<String, Object>> correlatedObject = new ArrayList<>();
+        Response response = new Response();
+        Map mergeeDBMap = new HashMap<String, Object>();
+        Map requestMap = userRequest.getRequest();
+        String mergeeId = (String)requestMap.get(JsonKey.FROM_ACCOUNT_ID);
+        String mergerId = (String)requestMap.get(JsonKey.TO_ACCOUNT_ID);
+        User mergee = userService.getUserById(mergeeId);
+        if(!mergee.getIsDeleted()) {
```
[the remark this answers is not in the document]
Comment (Hari-stackroute) [reply actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java]: created method prepareMergeeAccountData

File: actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+            userRequest.put(JsonKey.USER_MERGEE_ACCOUNT, mergeeDBMap);
+
+            //update the merger-course details in cassandra & ES
+            String mergerCourseResponse = updateMergerCourseDetails(requestMap);
+
+            if(mergerCourseResponse.equalsIgnoreCase(JsonKey.SUCCESS)) {
+                Response mergeeResponse = getUserDao().updateUser(mergeeDBMap);
+                String mergeeResponseStr = (String) mergeeResponse.get(JsonKey.RESPONSE);
+                ProjectLogger.log("UserMergeActor:UserMergeActor: mergeeResponseStr = " + mergeeResponseStr);
```
[the remark this answers is not in the document]
Comment (Hari-stackroute) [reply actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java]: Done

File: actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+                            JsonKey.FROM_ACCOUNT_ID),
+                    ResponseCode.CLIENT_ERROR.getResponseCode());
+        }
+
+
+    }
+
+    private void mergeUserDetailsToEs(Request userRequest) {
+        userRequest.setOperation(ActorOperations.MERGE_USER_TO_ELASTIC.getValue());
+        ProjectLogger.log(String.format("%s:%s:Trigger sync of user details to ES with user updated userMap %s", this.getClass().getSimpleName(), "mergeUserDetailsToEs"),LoggerEnum.INFO.name());
```
[the remark this answers is not in the document]
Comment (Hari-stackroute) [reply actors/user/src/main/java/org/sunbird/user/actors/UserMergeActor.java]: Ok removed and followed the normal way.

**Pull request merged**